### PR TITLE
Login/facebook

### PR DESCRIPTION
**Pull request by takayukioda**
Implement facebook login & refactor OAuthProvider
@shunjikonishi please review, thanks.

### CODE REVIEW

**Comment by takayukioda** [comment]
@shunjikonishi 
Sorry, I realized that I forgot to fix #4. Will fix after this review.

File: app/services/oauth2/AccessRequest.scala [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  protected val baseParams: Map[String, String] = Map(
+    "client_id" -> clientId,
+    "redirect_uri" -> redirectUri,
+    "state" -> state,
+    "response_type" -> "code"
+  )
+
+  def params: Map[String, String] = baseParams ++ optionalParams
+
+  def params(moreParams: Map[String, String]): Map[String, String] = baseParams ++ optionalParams ++ moreParams
```
Comment (shunjikonishi) [inline app/services/oauth2/AccessRequest.scala]: You can unify these param methods.

```
def params(moreParams: Map[String, String] = Map.empty): Map[String, String]
```

File: app/services/oauth2/TokenRequest.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    "redirect_uri" -> redirectUri,
+    "state" -> state,
+    "response_type" -> "code"
+  )
+
+  def params: Map[String, String] = baseParams ++ optionalParams
+
+  def params(moreParams: Map[String, String]): Map[String, String] = baseParams ++ optionalParams ++ moreParams
+
+}
```
Comment (shunjikonishi) [inline app/services/oauth2/TokenRequest.scala:26]: Maybe this class is too redundant.
These val and def are not used.
- clientId
- clientSecret
- redirectUri
- params

**Comment by shunjikonishi** [comment]
@da0shi I added some comments. fix me.

And more,
It is better to use trait instead of abstract class.

**Comment by shunjikonishi** [comment]
fix me

**Comment by takayukioda** [comment]
@shunjikonishi 

> And more,
> It is better to use trait instead of abstract class.

Is this talking about only TokenRequest or include AccessRequest?

**Comment by shunjikonishi** [comment]
Both. AccessRequest also can change to trait.

**Comment by takayukioda** [comment]
@shunjikonishi both AccessRequest and TokenRequest are now trait. please review.

**Comment by shunjikonishi** [comment]
@da0shi fix me

my intention is like that.

``` scala
trait A {
val f1: String
val f2: String
}

class B(val f1: String, val f2: String) extends A
```

**Comment by takayukioda** [comment]
Update implementation for trait. I hope this is what you meant.
@shunjikonishi please review, thanks.

**Comment by shunjikonishi** [comment]
LGTM

**Pull request merged**